snuba/datasets/groups.py

```diff
@@ -0,0 +1,124 @@
+                        table_alias=self.EVENTS_ALIAS,
+                        column="project_id"),
+                ),
+                JoinCondition(
+                    left=JoinConditionExpression(
+                        table_alias=self.GROUPS_ALIAS,
+                        column="id"),
+                    right=JoinConditionExpression(
+                        table_alias=self.EVENTS_ALIAS,
+                        column="group_id"),
+                ),
+            ],
+            join_type=JoinType.LEFT,
+        )
+
+        schema = JoinedSchema(join_structure)
+        dataset_schemas = DatasetSchemas(
+            read_schema=schema,
+            write_schema=None,
+        )
+        super().__init__(
+            dataset_schemas=dataset_schemas,
+            time_group_columns={
+                'events.time': 'events.timestamp',
+            },
+        )
+
+    def default_conditions(self):
+        return [
+            ('events.deleted', '=', 0),
+            ('groups.record_deleted', '=', 0),
+        ]
+
+    def column_expr(self, column_name, body, table_alias: str=""):
+        # Eventually joined dataset should not be represented by the same abstraction
+        # as joinable datasets. That will be easier through the TableStorage abstraciton.
+        # Thus, as of now, receiving a table_alias here is not supported.
+        assert table_alias == "", \
+            "Groups dataset cannot be referenced with table aliases. Alias provided {table_alias}"
+
+        match = self.QUALIFIED_COLUMN_REGEX.match(column_name)
+        if not match:
+            # anything that is not prefixed with a table alias is simply
+            # escaped and returned. It could be a literal.
+            return super().column_expr(column_name, body, table_alias)
+        else:
+            table_alias = match[1]
+            simple_column_name = match[2]
+            if table_alias == self.GROUPS_ALIAS:
+                return self.__grouped_message.column_expr(simple_column_name, body, table_alias)
+            elif table_alias == self.EVENTS_ALIAS:
+                return self.__events.column_expr(simple_column_name, body, table_alias)
+            else:
+                # This is probably an error condition. To keep consistency with the behavior
+                # in existing datasets, we let Clickhouse figure it out.
+                return super().column_expr(simple_column_name, body, table_alias)
+
+    def get_extensions(self) -> Mapping[str, QueryExtension]:
+        return {
+            'performance': PerformanceExtension(),
+            'project': ProjectExtension(),
+            'timeseries': TimeSeriesExtension(
+                default_granularity=3600,
+                default_window=timedelta(days=5),
+                timestamp_column='events.timestamp',
+            ),
+        }
+
+    def get_prewhere_keys(self) -> Sequence[str]:
+        return ['events.event_id', 'events.project_id']
```

Comment:
What's the thought process behind the prewhere keys for this dataset? (This isn't a leading question, I still don't really have a good handle on what good selection criteria are, I'm mostly curious.)

Reply:
There was basically no thought process. I think I should change it at least to include all the events column. But finally I guess we will have to revisit how to optimize the query in a world with joins.

Reply:
Makes sense, thanks for clarification. It'd probably be helpful to comment this kind of thing in the future when it's being originally written to make sure it doesn't slip through by accident (not that's it's going to cause any significant problem, just that it's easily missed and might be misinterpreted as intentional later.)